### PR TITLE
Allow for Union-typed memory aliasing

### PR DESCRIPTION
**Pull request by ansley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58911
* #58673
* #57014
* #57013
* __->__ #57012
* #57011
* #53180

Differential Revision: [D28572814](https://our.internmc.facebook.com/intern/diff/D28572814)

### CODE REVIEW

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

File: torch/csrc/jit/ir/alias_analysis.cpp
Comment (eellison) [inline torch/csrc/jit/ir/alias_analysis.cpp:93]: In the previous implementation, we were returning a vector of types here? why was that changed ?

**Review by eellison — COMMENTED** [review]
The last version of this I reviewed looked pretty close, but this looks very different now. Could we bring the implementation closer to how it was ? In particular, getMutableType should be returning a vector of mutable types as it was in the last version

File: torch/csrc/jit/ir/alias_analysis.cpp
Comment (ansley) [reply torch/csrc/jit/ir/alias_analysis.cpp:27]: Same semantics, but imo the original wording could be improved. I had to read that first sentence several times to picture it in my head. I don't mind keeping the original if I'm the only one who has problems with it, though :)

Comment (eellison) [reply torch/csrc/jit/ir/alias_analysis.cpp:27]: The semantics are a bit different, as the `Tuple[None, Tensor]` and `Tuple[List[int], None]` case show

File: torch/csrc/jit/ir/alias_analysis.cpp
Comment (ansley) [reply torch/csrc/jit/ir/alias_analysis.cpp:93]: In the current implementation, the signature is the same as it is here--`c10::optional<TypePtr> getMutableTypeImpl(const TypePtr& type)`. By not changing the signature, I was able to keep more of the original logic. Also, in most cases, we won't be returning multiple elements in the vector. Wouldn't returning a vector just cause lots of extra `if (vec.size() == ...` checks?

Comment (eellison) [reply torch/csrc/jit/ir/alias_analysis.cpp:93]: I think some of those changes are needed, like here https://github.com/pytorch/pytorch/blob/6dea82a8edce728b8c4850b884bbb976a1a59c18/torch/csrc/jit/ir/alias_analysis.cpp#L1153. 

If we refine a `Union[List[int], Tensor]` to `Tensor` we want the refined Tensor Value* to point to the Union Value*, but that won't occur right now because they will be returning different Mutable Types. 

`getMutableType` could have always been `mapTypeToTypeSetWhichItCanAlias`, however because the only form of Unions previously we're `Union[T, None]` we always knew that the typeSet was size 1

File: torch/csrc/jit/ir/alias_analysis.cpp
Comment (ansley) [reply torch/csrc/jit/ir/alias_analysis.cpp:27]: Is the issue here with "This function follows a similar logic to `unifyTypes` because..."? That line is from the original wording. It was confusing to me when I was trying to understand the alias analysis pass (and it sounds like you're saying it's downright misleading), so maybe we should take it out altogether

File: torch/csrc/jit/frontend/ir_emitter.cpp
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: Fixed in previous PR

File: test/jit/test_union.py
Comment (ansley) [reply test/jit/test_union.py]: I think this was from when I didn't understand `torch.jit.isinstance` 🥲

File: torch/csrc/jit/ir/alias_analysis.cpp
Comment (ansley) [reply torch/csrc/jit/ir/alias_analysis.cpp:93]: Resolved offline: my current implementation isn't wrong, but we will change `getMutableType` to return a `std::vector<TypePtr>` since that will make the API more intuitive

File: torch/csrc/jit/ir/alias_analysis.cpp
Comment (eellison) [inline torch/csrc/jit/ir/alias_analysis.cpp:113]: nit: you don't need to wrap it in `c10::optional<TypePtr>`

**Review by eellison — COMMENTED** [review]
Would it be possible to separate out the non-semantic changes? I think 80% of this review is non-semantic. I think separating that out would make iterating on making changes and reviewing much quicker, because the non-semantic changes do not require as close attention for reviewing.

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

File: torch/csrc/jit/passes/utils/memory_dag.cpp
Comment (ansley) [reply torch/csrc/jit/passes/utils/memory_dag.cpp:196]: The name of the variable has been changed from `wildcardElement` to `wildcardElement_List` to denote that it is no longer a single element but rather a list. I'm reverting the rest of the changes

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Review by eellison — COMMENTED** [review]
Looks great !! Couple of small comments, we discussed offline the 1-1 mapping of union wildcards to elements i think with that this is ready to go

File: torch/csrc/jit/ir/alias_analysis.cpp
Comment (eellison) [inline torch/csrc/jit/ir/alias_analysis.cpp:1271]: nit: I would add an assertion that the type is a Union type here. I dont know if we have Optional[Union] right now (eek) so maybe this is a todo after we remove optionals

File: torch/csrc/jit/ir/alias_analysis.cpp
Comment (eellison) [inline torch/csrc/jit/ir/alias_analysis.cpp:1253]: `add_wildcard_to_contained_elems` is a micro-optimization for nodes like 

```
x = torch.tensor([2])
torch.cat([x, x, x, x])
```
Where we look if the list created is passed to a functional operator and then put x in the contained elements of the list instead of setting x to the wildcard set. Even if this parameter is `False`,  we still want to `pointUnionTypeElementToAllContainedTypes`. It maybe makes sense to rename the argument to something like `contained_elements_have_already_added` or something, i'm not sure.

File: torch/csrc/jit/ir/alias_analysis.cpp
Comment (eellison) [inline torch/csrc/jit/ir/alias_analysis.cpp:63]: nit: maybe add an assertion here that `mutable_types->size() != 0`

**Review by eellison — APPROVED** [review]
Great job 🚀 🚀 🚀  

The one comment about adding `add_wildcard_to_contained_elems` should be easily addressable without another review from me so this is good to go from my end provided the `TORCH_INTERNAL_ASSERT(e->pointsTo.empty())` is fine to remove - I think it should but would like to double check with @suo before landing

File: torch/csrc/jit/ir/alias_analysis.cpp
Comment (ansley) [reply torch/csrc/jit/ir/alias_analysis.cpp:1271]: It's not possible to have `Optional[Union]`! I changed that in the first PR to always be `Union[None, ...]`

File: torch/csrc/jit/passes/utils/memory_dag.cpp
Comment (ansley) [reply torch/csrc/jit/passes/utils/memory_dag.cpp:192]: This is what I want the Union wildcards to look like:

```
WILDCARD(Union[T1, T2, T3]) -----> WILDCARD(T1)
                           |
                            -----> WILDCARD(T2)
                           |
                            -----> WILDCARD(T3)
```

So, the wildcard of a Union should have empty `values`, but it should point to other wildcards, which means that `pointsTo` won't be empty. This was not a possibility before implementing Union, which is why the assert worked

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572814).

**Pull request closed without merging**
